### PR TITLE
luaPackages.tree-sitter-norg: init at 0.2.4-1

### PR DESCRIPTION
**Pull request by mrcjkb**
## Description of changes

This will soon be a dependency of neorg.
It is also a POC for packaging tree-sitter parsers with luarocks.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by mrcjkb** [comment]
I also [have tree-sitter-haskell working](https://github.com/mrcjkb/neotest-haskell/pull/162) in one of my personal projects :tada:

**Review by GaetanLepage — APPROVED** [review]
This is very cool !

LGTM

**Comment by mrcjkb** [comment]
Result of `nixpkgs-review pr 309935` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>lua51Packages.luarocks-build-treesitter-parser</li>
    <li>lua51Packages.tree-sitter-norg</li>
    <li>lua52Packages.luarocks-build-treesitter-parser</li>
    <li>lua52Packages.tree-sitter-norg</li>
    <li>lua53Packages.luarocks-build-treesitter-parser</li>
    <li>lua53Packages.tree-sitter-norg</li>
    <li>lua54Packages.luarocks-build-treesitter-parser</li>
    <li>lua54Packages.tree-sitter-norg</li>
    <li>luajitPackages.luarocks-build-treesitter-parser</li>
    <li>luajitPackages.tree-sitter-norg</li>
  </ul>
</details>

**Comment by teto** [comment]
seems like it needs fixing for darwin (or mark it broken). Hopefully this doesn't get picked up as a vimplugin because it's going to be confusing for users which grammar to install.

**Comment by mrcjkb** [comment]
> seems like it needs fixing for darwin (or mark it broken).

I'll look into fixing it later. I think we need to use the tree-sitter CLI and clang on Darwin.

> Hopefully this doesn't get picked up as a vimplugin because it's going to be confusing for users which grammar to install.

I haven't added a vimPlugin for it.
nvim-treesitter have dropped tree-sitter-norg on `main` though 🤷

**Comment by teto** [comment]
> nvim-treesitter have dropped tree-sitter-norg on main though 🤷

and this is why the approach you take is the best !

> I'll look into fixing it later. I think we need to use the tree-sitter CLI and clang on Darwin.

sounds not easy. I am fine with merging it as broken for darwin. You do a lot already, maybe let darwin users fix it ?

**Comment by mrcjkb** [comment]
> sounds not easy. I am fine with merging it as broken for darwin. You do a lot already, maybe let darwin users fix it ?

`luarocks-build-treesitter-parser` will try to use the new `tree-sitter build` command if it detects a tree-sitter CLI installation,
so it shouldn't be too hard to get it working.

I've just pushed a change that [ought to fix it for darwin](https://github.com/nvim-neorocks/luarocks-build-treesitter-parser/actions/runs/8870583640/job/24352534463) :crossed_fingers:

**Comment by mrcjkb** [comment]
🤔 no idea how it works on Darwin in the other luarocks-build-treesitter-parser workflow, but not here. 
Guess I'll have to mark it as broken for now...

File: pkgs/development/lua-modules/overrides.nix [hunk 1]
```diff
@@ -725,6 +727,20 @@ in
 
   });
 
+  tree-sitter-norg = prev.tree-sitter-norg.overrideAttrs (oa: {
+    nativeBuildInputs = let
+      # HACK: luarocks-nix doesn't pick up rockspec build dependencies,
```
Comment (teto) [inline pkgs/development/lua-modules/overrides.nix:732]: it will after staging-next is merged

**Pull request merged**